### PR TITLE
Add numpy-style dtypes to Variable factories.

### PR DESCRIPTION
**Pull request by gchanan**
1) Add numpy-style dtypes corresponding to torch tensor types.  These are:
torch.float16, torch.float32, torch.float64, torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64
as well as torch.cuda, torch.sparse, and torch.cuda.sparse equivalents.

2) Adds "legacy" names for the above dtypes that correspond more closely to existing tensor names.  These are:
torch.half, torch.float, torch.double, torch.short, torch.int, torch.long.
torch.byte and torch.char don't exist because they either don't match numpy semantics or differ on different architectures.

3) Adds a "dtype" parameter to Variable factories (e.g. zeros, ones) that allows the user to specify the type without changing the default tensor type.

4) Adds a "dtype" getter to Variables that return the canonical dtype from 1)

This PR is missing the following useful features that should be added in the future:
A) We only add the "dtype" parameter to auto-generated factories; hand-written factories like in tensor_new.cpp don't support this yet.

B) We don't allow type conversions to use dtypes; that should be added to type(param) or a new function.

C) We don't yet have a "device" parameter for these factories; right now, they will only create Variables on the default device.

### CODE REVIEW

**Comment by gchanan** [comment]
The issue here appears to be if you compile with NO_CUDA=1, you don't define the cuda dtypes, even though as of now you get the torch.cuda.XTensor types.  I believe https://github.com/pytorch/pytorch/pull/5225 has the same issue, although maybe it isn't exposed by the existing tests (we should test it, though).

CC @colesbury.

**Comment by ezyang** [comment]
Looks nice. A few questions:

1. The intent is that once Variable-Tensor merge, then the Variable factories will replace `torch.zeros`, is that right?
2. Suppose I create a Variable using `_VariableFunctions.zeros`. What is the idiomatic way of creating a tensor that also requires gradient? Will we add a `requires_grad` flag to these functions? It kind of looks like `gen_python_functions.py` is handling this case but I don't see tests exercising it.

**Comment by gchanan** [comment]
1. Yes.
2. That already exists.  You are correct that there aren't tests that exercise both simultaneously though.

File: torch/csrc/autograd/python_variable.cpp [hunk 1]
```diff
@@ -431,6 +431,14 @@ PyObject *THPVariable_is_sparse(THPVariable *self)
   END_HANDLE_TH_ERRORS
 }
 
+PyObject *THPVariable_dtype(THPVariable *self)
+{
+  HANDLE_TH_ERRORS
+  auto& self_ = self->cdata;
+  return torch::autograd::utils::wrap((THPDtype*)torch::getDtype(self_.type()));
```
Comment (colesbury) [inline torch/csrc/autograd/python_variable.cpp]: The `THPDtype*` cast looks redundant.

File: torch/csrc/DynamicTypes.cpp [hunk 2]
```diff
@@ -100,6 +105,14 @@ at::Type& getATenType(PyTypeObject* type)
   throw std::invalid_argument("unsupported Python type");
 }
 
+THPDtype* getDtype(at::Type& type) {
+  auto it = attype_to_dtype.find(&type);
+  if (it != attype_to_dtype.end()) {
+    return it->second;
+  }
+  throw std::invalid_argument("unsupported at::Type");
```
Comment (colesbury) [inline torch/csrc/DynamicTypes.cpp]: Including the name of the function will help debugging in case this is actually hit. e.g.:

```
getDtype(): unsupported at::Type
```

File: torch/csrc/utils/python_arg_parser.h [hunk 3]
```diff
@@ -218,6 +223,22 @@ inline std::vector<int64_t> PythonArgs::intlistWithDefault(int i, std::vector<in
...
+  if (!THPDefaultATenType) {
+    throw std::runtime_error("THPDefaultATenType not initialized");
+  }
+  return *torch::autograd::VariableType::getType(*THPDefaultATenType);
+}
+
+inline const at::Type& PythonArgs::dtype(int i) {
+  if (!args[i]) return default_type();
+  PyObject *arg = args[i];
+  if (!THPDtype_Check(arg)) {
```
Comment (colesbury) [inline torch/csrc/utils/python_arg_parser.h]: Don't double-check the argument type. It's already checked by the parser.

File: torch/csrc/Dtype.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  PyObject_HEAD
+  at::Type *cdata;
+  const char* name;
+  bool is_cuda;
+  bool is_sparse;
+};
+
+extern PyObject *THPDtypeClass;
+
+#define THPDtype_Check(obj) ((PyObject*)Py_TYPE(obj) == THPDtypeClass)
```
Comment (colesbury) [inline torch/csrc/Dtype.h]: Prefer inline functions to macros.

**Comment by apaszke** [comment]
NB: I don't think that creating a tensor using one of those factories that also requires gradient is an often use case

File: test/test_torch.py [hunk 5]
```diff
@@ -1079,6 +1079,25 @@ def test_ones_like_multiple_device(self):
...
+                       torch.cuda.float16, torch.cuda.float32, torch.cuda.float64]
+        dtypes = cpu_dtypes + (cuda_dtypes if torch.cuda.is_available() else [])
+
+        for dtype in dtypes:
+            # no ops on float16 currently, cuda.float16 doesn't work on windows
+            if dtype != torch.float16 and not (dtype == torch.cuda.float16 and IS_WINDOWS):
+                out = torch._C._VariableFunctions.eye(3, 3, dtype=dtype)
+                self.assertEqual(out.dtype, dtype)
+            self.assertEqual(dtype in cuda_dtypes, dtype.is_cuda)
+            self.assertFalse(dtype.is_sparse)
```
Comment (apaszke) [inline test/test_torch.py]: Can we deduplicate this test into a generic thing, and call that from `test_sparse`?

File: torch/csrc/Dtype.h [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+PyObject * THPDtype_New(at::Type* cdata, const std::string& name, bool is_cuda, bool is_sparse);
+
+#ifdef _THP_CORE
```
Comment (colesbury) [inline torch/csrc/Dtype.h]: Don't do this stuff. The THP.h/_THP_CORE stuff is a mess. Peter's torch.h should be the preferred C++ interface for PYTorhc.

File: torch/csrc/utils/tensor_dtypes.cpp [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+#include "tensor_dtypes.h"
+#include "torch/csrc/Dtype.h"
+#include "torch/csrc/DynamicTypes.h"
+#include "torch/csrc/Exceptions.h"
+#include "torch/csrc/autograd/generated/VariableType.h"
+#include "torch/csrc/utils/tensor_types.h"
+
+namespace torch { namespace utils {
+
+static std::pair<std::string, std::string> getDtypeNames(at::ScalarType scalarType) {
```
Comment (apaszke) [inline torch/csrc/utils/tensor_dtypes.cpp]: Is there a reason why we don't want to make those `const char*`? Remember that this will create copies of both strings, irrespectively of what you'll be doing with them (not a big deal if this is only used in initialization though).

File: torch/csrc/utils/tensor_dtypes.cpp [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    default:
+      throw std::runtime_error("Unimplemented scalar type");
+  }
+}
+
+void initializeDtypes() {
+  auto torch_module = THPObjectPtr(PyImport_ImportModule("torch"));
+  auto cuda_module = THPObjectPtr(PyImport_ImportModule("torch.cuda"));
+  auto sparse_module = THPObjectPtr(PyImport_ImportModule("torch.sparse"));
+  auto cuda_sparse_module = THPObjectPtr(PyImport_ImportModule("torch.cuda.sparse"));
```
Comment (colesbury) [inline torch/csrc/utils/tensor_dtypes.cpp]: You should check for null (and throw `python_error()`) after these calls.

File: torch/csrc/utils/tensor_dtypes.cpp [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+      default: throw std::runtime_error("Unimplemented backend");
+    }
+    auto baseType = context.type_registry[static_cast<int>(backend)][static_cast<int>(scalarType)].get();
+    auto type = baseType ? torch::autograd::VariableType::getType(*baseType) : nullptr;
+    std::string name = std::string(PyModule_GetName(module)) + '.' + primary_name;
+    THPDtype *dtype = (THPDtype*)THPDtype_New(type, name, is_cuda, is_sparse);
+    if (type) {
+      torch::registerDtypeObject(dtype, *type);
+    }
+    PyModule_AddObject(module, primary_name.c_str(), (PyObject*)dtype);
```
Comment (colesbury) [inline torch/csrc/utils/tensor_dtypes.cpp]: Check the return value (`int`) for errors

File: torch/csrc/DynamicTypes.h [hunk 10]
```diff
@@ -19,10 +20,13 @@ void registerStoragePyTypeObject(
     PyTypeObject *pytype, const std::string& name,
     bool is_cuda, bool is_sparse);
 
+void registerDtypeObject(THPDtype *dtype, at::Type& type);
+
 PyObject* createPyObject(const at::Tensor& tensor);
 PyObject* createPyObject(const at::Storage& storage);
 PyTypeObject* getPyTypeObject(const at::Tensor& tensor);
 at::Type& getATenType(PyTypeObject* type);
+THPDtype* getDtype(at::Type& type);
```
Comment (colesbury) [inline torch/csrc/DynamicTypes.h]: `const at::Type&`?

File: torch/csrc/utils/tensor_dtypes.cpp (hunk 9 above)
Comment (colesbury) [reply torch/csrc/utils/tensor_dtypes.cpp]: Also, I think you need to `Py_INCREF` dtype here.

File: torch/csrc/utils/tensor_dtypes.cpp [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    if (legacy_name != "") {
+      PyModule_AddObject(module, legacy_name.c_str(), (PyObject*)dtype);
```
Comment (colesbury) [inline torch/csrc/utils/tensor_dtypes.cpp]: and `Py_INCREF` here

File: torch/csrc/utils/tensor_dtypes.cpp [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+      }
+      case at::kSparseCUDA: {
+        module = cuda_sparse_module.get();
+        is_cuda = true;
+        is_sparse = true;
+        break;
+      }
+      default: throw std::runtime_error("Unimplemented backend");
+    }
+    auto baseType = context.type_registry[static_cast<int>(backend)][static_cast<int>(scalarType)].get();
```
Comment (colesbury) [inline torch/csrc/utils/tensor_dtypes.cpp]: Might be worth a comment as to why you're not using `Context::getType(...)` here.

File: tools/autograd/templates/python_torch_functions_dispatch.h [hunk 13]
```diff
@@ -29,6 +29,23 @@ static at::Type& default_type() {
...
+  std::call_once(once, []() {
+    auto module = THPObjectPtr(PyImport_ImportModule("torch.cuda"));
+    if (!module) throw python_error();
+    auto res = THPObjectPtr(PyObject_CallMethod(module.get(), "_lazy_init", ""));
+    if (!res) throw python_error();
+  });
+}
+
+static const at::Type& maybe_initialize_cuda(const at::Type &type) {
+  if(type.is_cuda()) {
```
Comment (colesbury) [inline tools/autograd/templates/python_torch_functions_dispatch.h]: `if (`

File: test/test_torch.py (hunk 5 above)
Comment (gchanan) [reply test/test_torch.py]: I think so?  I'm not that familiar with test_sparse, but it seems possible.

File: torch/csrc/utils/tensor_dtypes.cpp (hunk 7 above)
Comment (gchanan) [reply torch/csrc/utils/tensor_dtypes.cpp]: I just did it that way because it's only in initialization and std::strings are nicer to use.  I can change it if you want.

File: tools/autograd/templates/python_torch_functions_dispatch.h [hunk 14]
```diff
@@ -29,6 +29,23 @@ static at::Type& default_type() {
...
+  static std::once_flag once;
+  std::call_once(once, []() {
+    auto module = THPObjectPtr(PyImport_ImportModule("torch.cuda"));
+    if (!module) throw python_error();
+    auto res = THPObjectPtr(PyObject_CallMethod(module.get(), "_lazy_init", ""));
+    if (!res) throw python_error();
+  });
+}
+
+static const at::Type& maybe_initialize_cuda(const at::Type &type) {
```
[the remark this answers is not in the document]
Comment (gchanan) [reply tools/autograd/templates/python_torch_functions_dispatch.h]: because at one point, I was doing the initialize + return inline, i.e.
`
maybe_initialize_cuda(type).foobar(...)
`

but this didn't jive well with how the codegen was written, so I removed it.  I'll remove the return type.

File: torch/csrc/utils/python_arg_parser.h (hunk 3 above)
Comment (gchanan) [reply torch/csrc/utils/python_arg_parser.h]: sure; fwiw this seems to be done often in these functions (see e.g. Generator).

File: torch/csrc/Dtype.h [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+
+struct THPDtype {
+  PyObject_HEAD
+  at::Type *cdata;
+  char name[DTYPE_NAME_LEN + 1];
+  bool is_cuda;
+  bool is_sparse;
+};
+
+extern PyObject *THPDtypeClass;
```
Comment (colesbury) [inline torch/csrc/Dtype.h]: This seems unnecessary to me. Use `extern PyTypeObject* THPDtypeType` instead. Then you don't have to worry about initializing it.

The reason we use the separate class/Type pattern in other places is because we subclass `_VariableBase` with `Variable` in Python, but want to construct the subclass from C++. Since there's no subclass of THPDtypeType here it doesn't matter.

File: torch/csrc/utils/tensor_dtypes.cpp [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+    // use this rather than context.getType() because getType throws exceptions.
+    auto baseType = context.type_registry[static_cast<int>(backend)][static_cast<int>(scalarType)].get();
+    auto type = baseType ? torch::autograd::VariableType::getType(*baseType) : nullptr;
+    std::string name = std::string(PyModule_GetName(module)) + '.' + primary_name;
+    PyObject *dtype = THPDtype_New(type, name, is_cuda, is_sparse);
+    if (type) {
+      torch::registerDtypeObject((THPDtype*)dtype, *type);
+    }
+    Py_INCREF(dtype);
+    if (PyModule_AddObject(module, primary_name.c_str(), (PyObject*)dtype) != 0) {
```
Comment (colesbury) [inline torch/csrc/utils/tensor_dtypes.cpp]: These `(PyObject*)` casts look unnecessary now

**Review by colesbury — APPROVED** [review]
See comments inline. Otherwise looks good to me

**Comment by gchanan** [comment]
CC @ezyang: this now supports Type as parameters to native functions, which will get turned into dtypes at the python level (assuming the function is bound).
